In [1]:
from LinAlgFuncs import *
from SymbolicCalcs import *

In [6]:
import sympy as sp

## 1 

Сначала найдём плоскость параллельную плоскости $-5x -y -2z=3$, проходящую через точку $(12, 13, -29)$: 
$\beta = -5x -y -2z = -5*12-13-2*-29 = -15$;  
Теперь найдём точку пересечения плоскости $\beta$ и прямой $x = -4t + 17, y = 4t, z = -5t -9$: 
$-5*(-4t+17) - (4t) -2*(-5t-9) = -15 \iff 26t = 52 \iff t = 2 \iff$ вторая точка = (9, 8, -19);

Итого уравнение прямой = $(12, 13, -29) + t(3, 5, -10)$

## 2 

In [18]:
A = Tensor([0,0,0])
B = Tensor([14, 0, 0])
C = Tensor([14,14,0])
D = Tensor([0,14,0])

A1 = Tensor([0,0,14])
B1 = Tensor([14, 0, 14])
C1 = Tensor([14,14,14])
D1 = Tensor([0,14,14])

E = (B+B1)*Fraction(1,2)
F = (4*B + 3*B1)*Fraction(1,7)
print(E, F)

[14,  0,  7] [14,  0,  6]


Угол между прямыми $AE$ и $D'F$ равен углу между векторами $AE$ и $D'F$

$$\angle(AE, D'F) = min(arccos(\frac{AE\times D'F}{|AE||D'F|}), 180 - arccos(\frac{AE\times D'F}{|AE||D'F|}))$$

In [19]:
ae = A-E
d1f = D1-F
dot(ae,d1f)*ExpressionPower(dot(ae,ae)*dot(d1f,d1f), Fraction(1,2))

Fraction(140, 1)Fraction(111720, 1)^Fraction(1, 2)

$\angle\left(AE, D'F\right) = min\left(arccos\left(\frac{140}{\sqrt{111720}}\right), 180 - arccos\left(\frac{140}{\sqrt{111720}}\right)\right) = min\left(arccos\left(\frac{\sqrt{570}}{57}\right), 180 - arccos\left(\frac{\sqrt{570}}{57}\right)\right)$

Чтобы найти расстояние между прямыми = $$\frac{\left|\left(\overline{AE}, \right)\right|}{\left|\left[\overline{AE}, \overline{D'F}\right]\right|}$$

In [25]:
dot(a,cross(b,c))

[0, 0, 0]

In [27]:
dot(ae, cross(d1f, E-F))/dot(cross(ae, d1f), cross(ae, d1f))

Fraction(-1, 470)

## 3 

In [ ]:
x = Variable("x")
A = tensor_from_iterable([
    [-12, 8, 5],
    [-12, 8, 6],
    [2, -2, -3]])

In [ ]:
B = (A - I(3)*x)
B

In [ ]:
print(det_by_minors(B))

$-x^3 -7x^2 -14x-8 = 0 \iff x^3 + 7x^2 + 14x + 8 = 0$. Заметим, что $-1^3 +7*-1^2 +14*-1 + 8 = 0 \Rightarrow x = -1$ - корень. 
$x^3 + 7x^2 + 14x + 8 = (x+1)*(x^2 + 6x + 8)$.  
Соответственно корни: $x = -1, \ x = \frac{-6 + \sqrt{36 - 4*8}}{2} = -2, \ x = \frac{-6 - \sqrt{36 - 4*8}}{2} = -4$ 

Чтобы найти собственные вектора, решим уравнение вида
$A\times v = \lambda * v \iff (A-\lambda * E)\times v = 0$

In [ ]:
v1 = solve_hsle(A - I(3)*-1)[0]
print(v1)
v2 = solve_hsle(A - I(3)*-2)[0]
print(v2)
v3 = solve_hsle(A - I(3)*-4)[0]
print(v3)

In [ ]:
rref(Tensor([v1, v2, v3]))

3 вектора линейно независимы $\Rightarrow$ они образуют в $\mathbb{R}^3$ базис $\Rightarrow \varphi$ - диагонализуем
базис, в котором $\varphi$ диагонален - $\left\{(-1, -2,  1),\left(\frac{1}{2},   0,   1\right),(1, 1, 0)\right\}$

При этом его матрица = $\begin{pmatrix}-1 & 0 &0\\
0 & -2 & 0\\
0 & 0 & -4\end{pmatrix}$

## 4

$Q(x_1, x_2, x_3) = 5x_1^2 + 5x_2^2 + 5x_3^2 -4x_1x_2 + 4x_1x_3 -4x_2x_3$

In [12]:
y = Variable("y")
A = tensor_from_iterable([
    [5, -2, 2],
    [-2, 5, -2],
    [2, -2, 5]
])

In [13]:
det_by_minors(A - I(3)*y)

(-1y^3 + 15y^2 + -63y + 81)

Характеристический многочлен A = $-y^3 + 15y^2 -63y + 81 = (y-3)(y^2- 12y + 27) = (y-3)^2(y-9)$

In [14]:
v1, v2 = solve_hsle(A - I(3)*3)
print(v1)
print(v2)
v3 = solve_hsle(A - I(3)*9)[0]
print(v3)

[1, 1, 0]
[-1,  0,  1]
[ 1, -1,  1]


In [15]:
e1 = Tensor([v1, v2, v3])
e = orthogonalize(e1)
print(e)
v1, v2, v3 = e

[[1, 1, 0], [-1/2,  1/2,    1], [ 1, -1,  1]]


In [ ]:
dot(v1, v1), dot(v2, v2), dot(v3, v3)

Отнормируем базис:

$(1,1,0)*\frac{1}{\sqrt{2}};\\
\left(-\frac{1}{2},\frac{1}{2},1\right)*\frac{\sqrt{2}}{\sqrt{3}};\\
(1, -1, 1)*\frac{1}{\sqrt{3}};$ 

$\left(\frac{1}{\sqrt{2}} , \frac{1}{\sqrt{2}} , 0 \right),\\
\left(-\frac{\sqrt{2}}{2\sqrt{3}}, \frac{\sqrt{2}}{2\sqrt{3}}, \frac{\sqrt{2}}{\sqrt{3}}\right),\\
\left(\frac{1}{\sqrt{3}}, -\frac{1}{\sqrt{3}}, \frac{1}{\sqrt{3}}\right)$

Тогда матрица перехода $ = \begin{pmatrix}\frac{1}{\sqrt{2}} & \frac{1}{\sqrt{2}} & 0\\
-\frac{\sqrt{2}}{2\sqrt{3}} & \frac{\sqrt{2}}{2\sqrt{3}} & \frac{\sqrt{2}}{\sqrt{3}}\\
\frac{1}{\sqrt{3}} & -\frac{1}{\sqrt{3}} & \frac{1}{\sqrt{3}}\end{pmatrix}$

Канонический вид квадратичной формы: $\begin{pmatrix}3& 0 & 0\\
0 & 3 & 0\\
0 & 0 & 9\end{pmatrix}$